### PR TITLE
server: add limiter config and reload mechanism

### PR DESCRIPTION
**Pull request by CabinfeverB**
Signed-off-by: Cabinfever_B <cabinfeveroier@gmail.com>

### What problem does this PR solve?

Issue Number: Ref #4666
**This PR should be merged after #4839**
**This PR should be merged after #4869 **
This PR is used to reload the limit config for limiter after leader resign or restart and do some integration tests

### What is changed and how it works?

update limiter config when reload presist config

```commit-message
update limiter config when reload presist config
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

### Release note

```release-note
None.
```

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -1475,6 +1506,14 @@ func (s *Server) reloadConfigFromKV() error {
 	return nil
 }
 
+func (s *Server) loadRateLimitConfig() {
```
Comment (rleungx) [inline server/server.go:1509]: Why do we need it?

File: server/config/service_middleware_config.go [hunk 2]
```diff
@@ -14,22 +14,30 @@
...
+	RateLimitConfig `json:"rate-limit"`
 }
 
 // NewServiceMiddlewareConfig returns a new service middleware config
 func NewServiceMiddlewareConfig() *ServiceMiddlewareConfig {
 	audit := AuditConfig{
 		EnableAudit: defaultEnableAuditMiddleware,
 	}
+	ratelimit := RateLimitConfig{
+		EnableRateLimit: defaultEnableRateLimitMiddleware,
```
Comment (rleungx) [inline server/config/service_middleware_config.go:36]: need make?

File: server/server.go (hunk 1 above)
Comment (CabinfeverB) [reply server/server.go:1509]: The configuration of the Limiter differs from that of EnableRateLimit. The value of EnableRateLimit is fetched from PersistOption on every request, but Limiter does not. So after resigning leader, we should reload the associated configuration.

File: server/api/service_middleware_test.go [hunk 3]
```diff
@@ -51,23 +51,27 @@ func (s *testServiceMiddlewareSuite) TearDownSuite(c *C) {
...
 		"audit.enable-audit": "false",
+		"enable-rate-limit":  "false",
 	}
 	postData, err = json.Marshal(ms)
 	c.Assert(err, IsNil)
 	c.Assert(tu.CheckPostJSON(testDialClient, addr, postData, tu.StatusOK(c)), IsNil)
 	sc = &config.ServiceMiddlewareConfig{}
 	c.Assert(tu.ReadGetJSON(c, testDialClient, addr, sc), IsNil)
 	c.Assert(sc.EnableAudit, Equals, false)
+	c.Assert(sc.EnableRateLimit, Equals, false)
```
Comment (AndreMouche) [inline server/api/service_middleware_test.go:74]: Could you please separate the test for rate-limit from `audit`?

File: server/server.go [hunk 4]
```diff
@@ -986,6 +990,27 @@ func (s *Server) SetAuditConfig(cfg config.AuditConfig) error {
...
+// GetRateLimitConfig gets the rate limit config information.
+func (s *Server) GetRateLimitConfig() *config.RateLimitConfig {
+	return s.serviceMiddlewarePersistOptions.GetRateLimitConfig().Clone()
+}
+
+// SetRateLimitConfig sets the rate limit config.
+func (s *Server) SetRateLimitConfig(cfg config.RateLimitConfig) error {
+	old := s.serviceMiddlewarePersistOptions.GetRateLimitConfig()
+	s.serviceMiddlewarePersistOptions.SetRateLimitConfig(&cfg)
+	if err := s.serviceMiddlewarePersistOptions.Persist(s.storage); err != nil {
```
Comment (AndreMouche) [inline server/server.go:1002]: Oh, Should we make this function threadsafe? Was it possible that two goroutines run this function, for example, should we avoid the following case:
1. go1 finished L1001 and update the cfg to false
2. go2 finished 1001 and update the cfg to true in cache
3. go2 persist the value with true (from cache) and return success
4. go1 persist the value with true(from cache) while it expect the value is false.
and more different cases may meet in the `Persist` with it not protect while read and write..

Comment (CabinfeverB) [reply server/server.go:1002]: This happens to all config that needs to `Persist`, and if optimization is needed, I think we can fix it together in another PR

File: server/config/service_middleware_config.go (hunk 2 above)
Comment (CabinfeverB) [reply server/config/service_middleware_config.go:36]: update

File: server/api/service_middleware_test.go (hunk 3 above)
Comment (CabinfeverB) [reply server/api/service_middleware_test.go]: > Meanwhile, could you please add some tests of updating the item of `limiter-config`?

Some tests will be added in #4843 after providering update API

File: tests/server/config/config_test.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+var _ = Suite(&testConfigPresistSuite{})
+
+type testConfigPresistSuite struct {
+	cleanup func()
+	cluster *tests.TestCluster
+}
+
+func (s *testConfigPresistSuite) SetUpSuite(c *C) {
+	ctx, cancel := context.WithCancel(context.Background())
+	server.EnableZap = true
```
Comment (rleungx) [inline tests/server/config/config_test.go:51]: This config is no longer used.

File: server/api/service_middleware_test.go (hunk 3 above)
Comment (CabinfeverB) [reply server/api/service_middleware_test.go]: > Meanwhile, could you please add some tests of updating the item of `limiter-config`?

I add a simple test, more test will in #4843 by specific API

File: server/api/service_middleware.go [hunk 6]
```diff
@@ -129,3 +132,24 @@ func (h *serviceMiddlewareHandler) updateAudit(config *config.ServiceMiddlewareC
...
 	return err
 }
+
+func (h *serviceMiddlewareHandler) updateRateLimit(config *config.ServiceMiddlewareConfig, key string, value interface{}) error {
+	data, err := json.Marshal(map[string]interface{}{key: value})
+	if err != nil {
+		return err
+	}
+
+	updated, found, err := mergeConfig(&config.RateLimitConfig, data)
```
Comment (rleungx) [inline server/api/service_middleware.go:142]: Do we need it?

File: server/api/service_middleware_test.go [hunk 7]
```diff
@@ -44,11 +45,11 @@ func (s *testServiceMiddlewareSuite) SetUpSuite(c *C) {
 	s.urlPrefix = fmt.Sprintf("%s%s/api/v1", addr, apiPrefix)
 }
 
-func (s *testServiceMiddlewareSuite) TearDownSuite(c *C) {
+func (s *testAuditMiddlewareSuite) TearDownSuite(c *C) {
 	s.cleanup()
 }
 
-func (s *testServiceMiddlewareSuite) TestConfigAudit(c *C) {
+func (s *testAuditMiddlewareSuite) TestConfigAuditSwitch(c *C) {
```
Comment (rleungx) [inline server/api/service_middleware_test.go:52]: Please add a test for the default value.

Comment (CabinfeverB) [reply server/api/service_middleware_test.go:52]: update

**Review by rleungx — APPROVED** [review]

**Review by AndreMouche — APPROVED** [review]
LGTM

**Pull request merged**